### PR TITLE
Spark streaming

### PR DESCRIPTION
**Pull request by sunheehnus**
Add streaming support

### CODE REVIEW

**Comment by itamarhaber** [comment]
:joy:

**Comment by dvirsky** [comment]
Thanks He. I'll go over it today or tomorrow!

**Comment by sunheehnus** [comment]
Thanks @dvirsky & @itamarhaber

File: src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+
+private class RedisReceiver[T: ClassTag](keys: Array[String],
+                                         storageLevel: StorageLevel,
+                                         redisConfig: RedisConfig,
+                                         streamType: Class[T])
+  extends Receiver[T](storageLevel) {
+
+  def onStart() {
+    val executorPool = ThreadUtils.newFixedThreadPool(keys.length, "BlockLists Streaming")
```
Comment (dvirsky) [inline src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala]: There's no need to do that. BLPOP supports popping from multiple keys, so a single loop is enough IMO.

`BLPOP key [key ...] timeout`

File: src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    }
+  }
+
+  def onStop() {
+  }
+
+  private class MessageHandler(conn: Jedis, key: String) extends Runnable {
+    def run() {
+      try {
+        while(!isStopped) {
```
Comment (dvirsky) [inline src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala]: if no data is arriving, the connector will be stuck forever in BLPOP and we will never stop.
This is why you should use timeouts for BLPOP...

File: src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala (hunk 1 above)
Comment (sunheehnus) [inline src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala]: I also thought that way, but I have a reason for this change:
   If the first part of the `keys` always have contents in them (eg: high rate input for those lists), the last part of the `keys` will be in a starvation sutiation.

File: src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala (hunk 2 above)
Comment (sunheehnus) [inline src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala]: `isStopped` is indeed a function of `Receiver` to check if receiver has been marked for stopping, so spark-streaming can use this to identify when the receiving of data should be stopped. So IMHO I don't think it has conflicts with BLPOP. :-)

Comment (dvirsky) [inline src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala]: so how can spark stop the receiver? there's no stopping condition other than isStopped

File: src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala (hunk 1 above)
Comment (dvirsky) [inline src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala]: ok, makes sense. I'm wondering though what will happen with many many keys. but leave it like this for now

File: src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala (hunk 2 above)
Comment (sunheehnus) [inline src/main/scala/com/redislabs/provider/redis/streaming/RedisInputDStream.scala]: Yes, you are right :-) ,  what do you suggest the timeouts for blpop or we just left it in the API?

**Pull request merged**